packages/server/src/target/build.rs

```diff
@@ -258,6 +262,68 @@ impl Server {
 		Ok(Some(output))
 	}
 
+	async fn detect_depth_overflow(&self, parent: &tg::build::Id) -> tg::Result<bool> {
+		let connection = self
+			.database
+			.connection(db::Priority::Low)
+			.await
+			.map_err(|source| tg::error!(!source, "failed to get a connection"))?;
+		let p = connection.p();
+
+		#[derive(serde::Deserialize, serde::Serialize)]
+		struct Row {
+			id: tg::build::Id,
+			depth: u64,
+		}
+		let statement = formatdoc!(
+			"
+				with recursive ancestors as (
+					select b.id, b.depth
+					from builds b
+					join build_children c on b.id = c.child
+					where c.child = {p}1
+
+					union all
+
+					select b.id, b.depth
+					from ancestors a
+					join build_children c on a.id = c.child
+					join builds b on c.build = b.id
+				)
+				select id, depth from ancestors;
+			"
+		);
+		let params = db::params![parent];
+		let ancestors = connection
+			.query_all_into::<Row>(statement, params)
+			.await
+			.map_err(|source| tg::error!(!source, "failed to execute the statement"))?;
+
+		let max_depth = ancestors.iter().map(|row| row.depth).max();
+		let ancestors = ancestors.iter().map(|row| row.id.clone()).collect_vec();
+		let ancestors = serde_json::to_string(&ancestors).unwrap();
+		if let Some(max_depth) = max_depth {
+			if max_depth >= 100 {
```

Comment:
`100` has been a good value for testing, but we need to bump this before submitting the PR